engine/runtime/src/main/java/org/enso/interpreter/runtime/library/dispatch/TypeOfNode.java

```diff
@@ -24,55 +24,135 @@
 import org.enso.interpreter.runtime.number.EnsoBigInteger;
 import org.enso.interpreter.runtime.warning.WithWarnings;
 
+/**
+ * Provides API (in form of {@code public} methods) for querying type information about a value.
+ * Contains non-{@code public} methods with implementation to handle such queries.
+ */
 @GenerateUncached
 public abstract class TypeOfNode extends Node {
   TypeOfNode() {}
 
-  public abstract Object execute(Object value);
+  /**
+   * Check whether given value has an Enso {@link Type}.
+   *
+   * @param value the value to check
+   * @return {@code true} if there is (at least one) type associated with the value
+   */
+  public final boolean hasType(Object value) {
+    return findTypeOrError(value) instanceof Type;
+  }
+
+  /**
+   * Finds (a primary) type associated with a given {@code value}. If no such type exists, an error
+   * value is returned that can normally flow thru the Enso interpreter without any conversions.
+   *
+   * @param value the value to check
+   * @return either {@link Type} of the value or {@link DataflowError} if there is no such type
+   */
+  public final Object findTypeOrError(Object value) {
+    try {
+      var types = executeTypes(value);
+      return types[0];
+    } catch (NonTypeResult plain) {
+      return plain.result;
+    }
+  }
+
+  /**
+   * Finds (a primary) type associated with a given {@code value} or returns {@code null}. Useful
+   * for internal checks in the Enso interpreter.
+   *
+   * @param value the value to check
+   * @return either Type of the value or {@code null} if there is no such type
+   */
+  public final Type findTypeOrNull(Object value) {
+    return findTypeOrError(value) instanceof Type type ? type : null;
+  }
+
+  /**
+   * Finds all types associated with a given {@code value} or returns {@code null}. It is guaranteed
+   * that the returned array is going to have at least one element, if it is non-{@code null}.
+   *
+   * @param value the value to check
+   * @return either types associated with the value or {@code null} if there is no such type
+   */
+  public final Type[] findAllTypes(Object value) {
```

Comment:
How do you like the `findXyzOrAbc` naming, guys? Should this be `findAllTypesOrNull` then? Thumbs up to rename to `findAllTypesOrNull`

Reply:
Renamed in 39ebae8